### PR TITLE
Add parsing of barcodes report

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
Add functionality to read the run-wise barcode metrics and the sample-wise barcoding metrics from the `barcodes.report.json`

### Added

- Constant class with the names of the fields in the metrics files (more DRY)
- Functions `_parse_sample_data` and `get_parsed_sample_metrics` in `cg/services/run_devices/pacbio/metrics_parser/utils.py` to parse the 'tables' (sample) section of the barcodes matrics file
- Calls to functions to parse barcodes metrics file
- `barcodes` and `samples` sections to the `PacBioMetrics` metrics model
- Fixtures

### Changed

- Updated any literals referring to metrics files fields with constants

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/constants/pacbio.py [hunk 1]
```diff
@@ -17,6 +17,15 @@ class PacBioDirsAndFiles:
     RUN_IS_VALID: str = "is_valid"
 
 
+class MetricsFileFields:
+    ATTRIBUTES: str = "attributes"
```
Comment (islean) [inline cg/constants/pacbio.py:21]: Just a minor thing and not anything you need to change but is there a reason for not considering this a class instantiation and foregoing the type hint? To me it is just as clear that this is a string as a custom object instantiation.

File: cg/constants/pacbio.py [hunk 2]
```diff
@@ -17,6 +17,15 @@ class PacBioDirsAndFiles:
...
+    COLUMNS: str = "columns"
+    ID: str = "id"
+    TABLES: str = "tables"
+    VALUE_ONE: str = "value"
+    VALUES_MANY: str = "values"
```
Comment (islean) [inline cg/constants/pacbio.py:26]: I think breaking the established pattern is more troublesome than potential typos. I would expect `VALUE_ONE = "value_one"` given how the rest is written.

```suggestion
    VALUE: str = "value"
    VALUES: str = "values"
```

File: cg/services/run_devices/pacbio/metrics_parser/utils.py [hunk 3]
```diff
@@ -32,12 +35,39 @@ def _parse_report_to_model(report_file: Path, data_model: Type[BaseMetrics]) ->
     parsed_json: dict = ReadFile.read_file[FileFormat.JSON](report_file)
     if data_model == SmrtlinkDatasetsMetrics:
         return data_model.model_validate(parsed_json[0], from_attributes=True)
-    metrics: list[dict[str, Any]] = parsed_json.get("attributes")
-    data: dict = {report_field["id"]: report_field["value"] for report_field in metrics}
+    metrics: list[dict[str, Any]] = parsed_json.get(MetricsFileFields.ATTRIBUTES)
+    data: dict = {
+        report_field["id"]: report_field[MetricsFileFields.VALUE_ONE] for report_field in metrics
```
Comment (islean) [inline cg/services/run_devices/pacbio/metrics_parser/utils.py:40]: Why do we always use the key "id" here? And if we have created the MetricsFileFields class, should that string not be in there?

File: cg/services/run_devices/pacbio/metrics_parser/utils.py [hunk 4]
```diff
@@ -32,12 +35,39 @@ def _parse_report_to_model(report_file: Path, data_model: Type[BaseMetrics]) ->
...
+    }
     return data_model.model_validate(data, from_attributes=True)
```
Comment (islean) [inline cg/services/run_devices/pacbio/metrics_parser/utils.py:42]: Does the "from_attributes" do anything here? If not, I think it adds confusion.

File: cg/services/run_devices/pacbio/metrics_parser/utils.py [hunk 5]
```diff
@@ -32,12 +35,39 @@ def _parse_report_to_model(report_file: Path, data_model: Type[BaseMetrics]) ->
...
+def _parse_sample_data(sample_data: list[dict[str, Any]]) -> list[SampleMetrics]:
+    """Parse all samples data into SampleMetrics given the sample section of the barcodes report."""
+    sample_metrics: list[SampleMetrics] = []
+    number_of_samples: int = len(sample_data[0].get(MetricsFileFields.VALUES_MANY))
+    for sample_idx in range(number_of_samples):
+        sample: dict = {}
+        for data_field in sample_data:
+            field_id: str = data_field.get(MetricsFileFields.ID)
+            sample[field_id] = data_field.get(MetricsFileFields.VALUES_MANY)[sample_idx]
+        sample_metrics.append(SampleMetrics.model_validate(sample, from_attributes=True))
```
Comment (islean) [inline cg/services/run_devices/pacbio/metrics_parser/utils.py:60]: Does the `from_attributes` do anything here?

File: cg/services/run_devices/pacbio/metrics_parser/utils.py [hunk 6]
```diff
@@ -32,12 +35,39 @@ def _parse_report_to_model(report_file: Path, data_model: Type[BaseMetrics]) ->
...
     return data_model.model_validate(data, from_attributes=True)
 
 
 def get_parsed_metrics_from_file_name(metrics_files: list[Path], file_name: str) -> BaseMetrics:
     report_file: Path = get_file_with_pattern_from_list(files=metrics_files, pattern=file_name)
     data_model: Type[BaseMetrics] = _get_data_model_from_pattern(pattern=file_name)
     return _parse_report_to_model(report_file=report_file, data_model=data_model)
+
+
+def _parse_sample_data(sample_data: list[dict[str, Any]]) -> list[SampleMetrics]:
```
Comment (islean) [inline cg/services/run_devices/pacbio/metrics_parser/utils.py:51]: I find this method and the whole parsing quite hard to read, and quite dependent on me having knowledge of how the file we are trying to parse is structured. With my brain being a bit mushy, would it be possible to use the csv module and some syntax similar to this?

image

**Review by islean — APPROVED** [review]
It probably does what we want but my brain is a bit too mushy to parse the parsing logic easily so take the review with a grain of salt. Is there a clearer way to write it?

File: cg/constants/pacbio.py (hunk 2 above)
Comment (diitaz93) [reply cg/constants/pacbio.py:26]: I was unsure of this one, thanks for the feedback, I will change it

File: cg/constants/pacbio.py (hunk 1 above)
Comment (diitaz93) [reply cg/constants/pacbio.py:21]: I think it helps in PyCharm when assigning this to any variable. I don't see any other benefit though

File: cg/services/run_devices/pacbio/metrics_parser/utils.py (hunk 4 above)
Comment (diitaz93) [reply cg/services/run_devices/pacbio/metrics_parser/utils.py:42]: This is to populate the class with the aliases, am I wrong? @islean

File: cg/services/run_devices/pacbio/metrics_parser/utils.py (hunk 5 above)
Comment (diitaz93) [reply cg/services/run_devices/pacbio/metrics_parser/utils.py:60]: see above

File: cg/services/run_devices/pacbio/metrics_parser/utils.py (hunk 6 above)
Comment (diitaz93) [reply cg/services/run_devices/pacbio/metrics_parser/utils.py:51]: I will leave it as it is and we can have a discussion in the development branch when more changes are in place

File: cg/services/run_devices/pacbio/metrics_parser/utils.py (hunk 4 above)
Comment (islean) [reply cg/services/run_devices/pacbio/metrics_parser/utils.py:42]: The aliases are used by default!  My understanding is that `from_attributes` is more for if you send another object with overlapping field names. So if you pass a database SQLAlchemy Base object to a Pydantic DTO, you can set `from_attributes = True` and use a `PydanticObject.model_validate(database_object, from_attributes=True)` immediately. See https://docs.pydantic.dev/latest/concepts/models/#arbitrary-class-instances.

Comment (diitaz93) [reply cg/services/run_devices/pacbio/metrics_parser/utils.py:43]: Ok, will remove, thanks!

**Pull request merged**